sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/ThrottleWithExternalResource.java

```diff
@@ -0,0 +1,409 @@
+              RedisEnqueuer<@NonNull T>,
+              RedisDequeuer<@NonNull T>,
+              RedisDecrementer,
+              RedisRefresher>
+          usingRedis(
+              URI uri,
+              String quotaIdentifier,
+              String queueKey,
+              Quota quota,
+              Coder<@NonNull T> coder)
+              throws Coder.NonDeterministicException {
+    return new ThrottleWithExternalResource<
+        @NonNull T,
+        RedisReporter,
+        RedisEnqueuer<@NonNull T>,
+        RedisDequeuer<@NonNull T>,
+        RedisDecrementer,
+        RedisRefresher>(
+        quota,
+        quotaIdentifier,
+        coder,
+        new RedisReporter(uri),
+        new RedisEnqueuer<>(uri, queueKey, coder),
+        new RedisDequeuer<>(uri, coder, queueKey),
+        new RedisDecrementer(uri, queueKey),
+        new RedisRefresher(uri, quota, quotaIdentifier));
+  }
+
+  private static final Duration THROTTLE_INTERVAL = Duration.standardSeconds(1L);
+
+  private final @NonNull Quota quota;
+  private final @NonNull String quotaIdentifier;
+  private final @NonNull Coder<@NonNull T> coder;
+  private final @NonNull ReporterT reporterT;
+  private final @NonNull EnqueuerT enqueuerT;
+  private final @NonNull DequeuerT dequeuerT;
+  private final @NonNull DecrementerT decrementerT;
+  private final @NonNull RefresherT refresherT;
+
+  ThrottleWithExternalResource(
+      @NonNull Quota quota,
+      @NonNull String quotaIdentifier,
+      @NonNull Coder<@NonNull T> coder,
+      @NonNull ReporterT reporterT,
+      @NonNull EnqueuerT enqueuerT,
+      @NonNull DequeuerT dequeuerT,
+      @NonNull DecrementerT decrementerT,
+      @NonNull RefresherT refresherT)
+      throws Coder.NonDeterministicException {
+    this.quotaIdentifier = quotaIdentifier;
+    this.reporterT = reporterT;
+    coder.verifyDeterministic();
+    checkArgument(!quotaIdentifier.isEmpty());
+    this.quota = quota;
+    this.coder = coder;
+    this.enqueuerT = enqueuerT;
+    this.dequeuerT = dequeuerT;
+    this.decrementerT = decrementerT;
+    this.refresherT = refresherT;
+  }
+
+  @Override
+  public Call.@NonNull Result<@NonNull T> expand(PCollection<@NonNull T> input) {
+    Pipeline pipeline = input.getPipeline();
+
+    // Refresh known quota to control the throttle rate.
+    Call.Result<Void> refreshResult =
+        pipeline
+            .apply("quota/impulse", PeriodicImpulse.create().withInterval(quota.getInterval()))
+            .apply("quota/refresh", getRefresher());
+
+    // Enqueue T elements.
+    Call.Result<Void> enqueuResult = input.apply("enqueue", getEnqueuer());
+
+    TupleTag<@NonNull T> outputTag = new TupleTag<@NonNull T>() {};
+    TupleTag<@NonNull ApiIOError> failureTag = new TupleTag<@NonNull ApiIOError>() {};
+
+    // Perform Throttle.
+    PCollectionTuple pct =
+        pipeline
+            .apply("throttle/impulse", PeriodicImpulse.create().withInterval(THROTTLE_INTERVAL))
```

Comment:
(probably the only important comment) In general, it's hard to imagine how could an endless periodicimpulse work in a batch pipeline

Reply:
The code comments above comment about this:
```
* <p>{@link ThrottleWithExternalResource} makes use of {@link PeriodicImpulse} as it needs to
 * coordinate three {@link PTransform}s concurrently. Usage of {@link ThrottleWithExternalResource}
 * should consider the impact of {@link PeriodicImpulse} on the pipeline.
```

In the public facing RequestResponseIO there will be a `withRedisBasedThrottle` or some method to be determined. This will impact on the pipeline i.e. batch to stream will be explicitly commented as a limitation. It's moreover emphasized that using this is optional and not to be relied on solely as a way to throttle a PCollection.